lib/active_merchant/billing/gateways/stripe_payment_intents.rb

```diff
@@ -478,30 +490,34 @@ def setup_future_usage(post, options = {})
       def add_billing_address_for_card_tokenization(post, options = {})
         return unless (billing = options[:billing_address] || options[:address])
 
-        post[:card][:address_city] = billing[:city] if billing[:city]
-        post[:card][:address_country] = billing[:country] if billing[:country]
-        post[:card][:address_line1] = billing[:address1] if billing[:address1]
-        post[:card][:address_line2] = billing[:address2] if billing[:address2]
-        post[:card][:address_zip] = billing[:zip] if billing[:zip]
-        post[:card][:address_state] = billing[:state] if billing[:state]
+        address = add_address({}, billing, options)
+        modify_address = address[:address].transform_keys { |k| k == :postal_code ? :address_zip : k.to_s.prepend('address_').to_sym }.merge!({ name: address[:name] })
+
+        post[:card].merge!(modify_address)
       end
 
-      def add_billing_address(post, options = {})
-        return unless billing = options[:billing_address] || options[:address]
+      def add_shipping_address(post, options = {})
+        return unless shipping = options[:shipping_address]
 
-        email = billing[:email] || options[:email]
+        post[:shipping] = add_address({}, shipping, options)
+        post[:shipping][:carrier] = (shipping[:carrier] || options[:shipping_carrier]) if shipping[:carrier] || options[:shipping_carrier]
+        post[:shipping][:tracking_number] = (shipping[:tracking_number] || options[:shipping_tracking_number]) if shipping[:tracking_number] || options[:shipping_tracking_number]
+      end
 
-        post[:billing_details] = {}
-        post[:billing_details][:address] = {}
-        post[:billing_details][:address][:city] = billing[:city] if billing[:city]
-        post[:billing_details][:address][:country] = billing[:country] if billing[:country]
-        post[:billing_details][:address][:line1] = billing[:address1] if billing[:address1]
-        post[:billing_details][:address][:line2] = billing[:address2] if billing[:address2]
-        post[:billing_details][:address][:postal_code] = billing[:zip] if billing[:zip]
-        post[:billing_details][:address][:state] = billing[:state] if billing[:state]
-        post[:billing_details][:email] = email if email
-        post[:billing_details][:name] = billing[:name] if billing[:name]
-        post[:billing_details][:phone] = billing[:phone] if billing[:phone]
+      def add_address(post, address, options)
```

Comment:
Why are we passing in `post` if we don't add to it here?

Reply:
I'll remove sending post